pkg/util/mount/mountpoints_cache_linux.go

```diff
@@ -0,0 +1,137 @@
+// +build linux
+
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mount
+
+import (
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"golang.org/x/sys/unix"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+// mountPointsCache implements cache of mountpoints on linux platform.
+// We use poll() on /proc/mounts to get notified about mount tree changes.
+type mountPointsCache struct {
+	sync.RWMutex
+	// Cache of mountpoints.
+	mountpoints []MountPoint
+	// Indicates cache is stale or not.
+	stale bool
+}
+
+func newMountPointsCache() *mountPointsCache {
+	cache := &mountPointsCache{
+		mountpoints: nil,
+		stale:       true,
+	}
+	return cache
+}
+
+// Get gets mountpoints cache if cache is fresh, otherwise calls `list()` to
+// list mountpoints. If `list()` succeeds, update cache and mark it fresh,
+// otherwise returns error.
+func (c *mountPointsCache) Get(list func() ([]MountPoint, error)) ([]MountPoint, error) {
+	c.RLock()
+	if !c.stale {
+		c.RUnlock()
+		return c.mountpoints, nil
+	}
+	c.RUnlock()
+	c.Lock()
+	defer c.Unlock()
+	// Double-checking locking criterion.
+	if !c.stale {
```

Comment:
Why do we need this? Once this function enters this section - the caller goroutine will have exclusive lock on code that could remove stale flag right?

Reply:
The first lock criterion is checked with reader lock not a exclusive lock. It's optimistic lock hint. We need to check the locking criterion again before proceeding the actual locking logic.

e.g. It's possible A and B acquired reader lock and found `c.stale` is true at the same time. Both of them will try to lock and proceed. If we don't checking locking criterion again, `list()` will be processed twice.

Reply:
Cool - I see that you just wanted to return early in second goroutine if first goroutine acquired lock and updated the cache.